[NgYaoDong] iP

NgYaoDong opened the pull request.

NgYaoDong committed:
Level-0

NgYaoDong committed:
Add Level-1: Echo

NgYaoDong committed:
Add Level-2: Add, List

NgYaoDong committed:
Add Level-3: Mark as Done + A-CodingStandard

NgYaoDong committed:
Add Level-4

Add support for tracking three types of tasks: ToDos, Events, Deadlines.

NgYaoDong committed:
Add Level-5

Add error handling

NgYaoDong committed:
Merge pull request #1 from NgYaoDong/branch-Level-5

Add Level-5

NgYaoDong committed:
Merge branch 'master' of https://github.com/NgYaoDong/ip

NgYaoDong committed:
Add A-Packages

Divide classes into multiple packages

NgYaoDong committed:
Merge pull request #2 from NgYaoDong/branch-A-Packages

Add A-Packages

NgYaoDong committed:
Merge branch 'branch-A-Packages'

NgYaoDong committed:
Merge branch 'master' of https://github.com/NgYaoDong/ip

NgYaoDong committed:
Add description of command usage

Add description, parameters and examples of all commands

NgYaoDong committed:
Add delete command

Add support for deleting tasks from the list.

rouvinerh commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        if (INPUT.isEmpty()) {
+            throw new DukeException("Exceed Charge....\n\t " +
```
I like that it provides guidance to the user for each command when an invalid input is entered.

rouvinerh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,41 @@
+        case "unmark":
+            return new UnmarkCommand(description);
+        case "todo":
+            //add todo
```
I think these comments can be removed since the functionality has been added.

rouvinerh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,38 @@
+
+import duke.ui.Ui;
+import duke.storage.Storage;
+import duke.command.Command;
```
Code uses a lot of classes and OOP features, which I think is good.

rouvinerh commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+public class DukeException extends Exception {
```
Good usage of try catch exceptions throughout the code.

rouvinerh reviewed (commented):
I like the way the bot is coded, plenty of OOP use and classees for each functionality.

TimothyLKM commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,38 @@
+import duke.ui.Ui;
+import duke.storage.Storage;
+import duke.command.Command;
+import duke.parser.Parser;
```
Too many classes, too much abstraction so is rather tedious to follow the code.

NgYaoDong committed:
Edit output of deleting task, and include delete task test case

NgYaoDong committed:
Add data saving feature

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.

NgYaoDong committed:
Merge branch 'branch-Level-6'

NgYaoDong committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/EventCommand.java

NgYaoDong committed:
Fix previous merge conflicts, and remove splitWords() method

NgYaoDong committed:
Change Task class to abstract class, and edit EXPECTED.TXT

NgYaoDong committed:
Add dates and times

Teach the chatbot how to understand dates and times.

NgYaoDong committed:
Add find command

Give users a way to find a task by searching for a keyword.

NgYaoDong committed:
Add JavaDoc comments

Add JavaDoc comments to the code.

NgYaoDong committed:
Merge pull request #3 from NgYaoDong/branch-Level-8

Add dates and times

NgYaoDong committed:
Merge branch 'master' of https://github.com/NgYaoDong/ip

NgYaoDong committed:
Merge branch 'master' into branch-Level-9

NgYaoDong committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java

NgYaoDong committed:
Merge pull request #4 from NgYaoDong/branch-Level-9

Add find command

NgYaoDong committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/parser/Parser.java

NgYaoDong committed:
Merge pull request #5 from NgYaoDong/branch-A-JavaDoc

Add JavaDoc comments

NgYaoDong committed:
Add a User Guide

Add a User Guide to the project in the specified way.

NgYaoDong committed:
Refine the User Guide

json-bot-boop commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException, IOException {
+        String[] details = INPUT.split("/by");
+        if (INPUT.isEmpty() || details.length != 2 || details[0].trim().isEmpty() || details[1].trim().isEmpty()) {
+            throw new DukeException("Exceed Charge....\n\t " +
+                    "OOPS!!! The description of a deadline task cannot be empty.\n\t " +
+                    "deadline: Adds a deadline task to task list.\n\t " +
+                    "Parameters: TASK /by DEADLINE (in date: yyyy-mm-dd time: HHmm format)\n\t " +
+                    "Example: deadline return book /by 2024-05-01 1800");
```
Try to avoid magic strings

json-bot-boop commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException, IOException {
+        String[] details = INPUT.split("/by");
+        if (INPUT.isEmpty() || details.length != 2 || details[0].trim().isEmpty() || details[1].trim().isEmpty()) {
```
Try to avoid complex expression

json-bot-boop commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+                    "Example: deadline return book /by 2024-05-01 1800");
+        } else {
+            String description = details[0].trim();
+            String by = details[1].trim();
```
Try to avoid magic numbers

json-bot-boop commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,76 @@
+        String[] details = INPUT.split("/from");
+        if (INPUT.isEmpty() || details.length != 2 || details[1].trim().isEmpty()) {
+            throw new DukeException(EXCEPTION);
+        } else {
+            String[] times = details[1].split("/to");
+            if (times.length != 2 || times[0].trim().isEmpty() || times[1].trim().isEmpty()) {
+                throw new DukeException(EXCEPTION);
+            }
+            String description = details[0].trim();
+            String from = times[0].trim();
+            String to = times[1].trim();
+            Task newEvent = new Event(description, from, to);
+            storage.addTask(newEvent.toDisk());
+            taskList.add(newEvent);
+            String msg = (taskList.size() > 1) ? "tasks" : "task";
+            ui.printMessage("Got it. I've added this task:\n\t   " + newEvent
+                    + "\n\t Now you have " + taskList.size() + " " + msg + " in the list.");
+        }
```
Maybe can consider removing the `else`

json-bot-boop commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     *
+     * @param FILEPATH The filepath of the saved text file containing the tasks.
+     */
+    public Storage(String FILEPATH) {
+        this.FILEPATH = FILEPATH;
+    }
```
Variable name in argument should not be capitalized

json-bot-boop commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,56 @@
+package duke.parser;
+
+import duke.DukeException;
+import duke.command.*;
```
Imported classes should always be listed explicitly.

json-bot-boop reviewed (commented):
Overall, the code is okay except some minor code quality and coding standard issues

NgYaoDong committed:
Refactor code to comply with coding standard

- Refactor to remove magic literals
- Add documentation
- Edit text-ui-test
